packages/wonder-blocks-core/components/unique-id-provider.md

```diff
@@ -0,0 +1,101 @@
+
+### mockOnFirstRender absent or false
+
+When no `mockOnFirstRender` is `false` (the default), the `children` prop is only called after the initial render. Each call provides the same identifier factory, meaning the same identifier gets returned. Try it below.
+
+```jsx
+const {Body, HeadingSmall} = require("@khanacademy/wonder-blocks-typography");
+const {Spring, Strut} = require("@khanacademy/wonder-blocks-layout");
+const Button = require("@khanacademy/wonder-blocks-button").default;
+
+let providerRef = null;
+
+const renders = [];
+const provider = (
+    <UniqueIDProvider ref={ref => providerRef = ref}>
+        {ids => {
+            renders.push(ids.id("my-unique-id"));
+            return (
+                <View>
+                    {renders.map((id,i) => (
+                        <Body key={i}>Render {i}: {id}</Body>
+                    ))}
+                </View>
+            );
+        }}
+    </UniqueIDProvider>
+);
+
+const onClick = () => {
+    if (providerRef) {
+        providerRef.forceUpdate();
+    }
+};
+
+<View>
+    <View style={{flexDirection: "row"}}><Button onClick={onClick}>Click Me to Rerender</Button><Spring /></View>
+    <Strut size={16} />
+    <HeadingSmall>The UniqueIDProvider:</HeadingSmall>
+    {provider}
+</View>
+```
+
+### mockOnFirstRender is true
+
+When specifying `mockOnFirstRender` to be `true`, the first render will use a mock identifier factory that doesn't guarantee identifier uniqueness.
+
+```jsx
+const {Body, HeadingSmall} = require("@khanacademy/wonder-blocks-typography");
+const {Spring, Strut} = require("@khanacademy/wonder-blocks-layout");
+
+let firstId = null;
+
+const children = (idf) => {
+    const id = idf.id("an-id");
+    firstId = firstId || id;
+    return (
+        <View>
+            <HeadingSmall>The initial render ID:</HeadingSmall>
+            {firstId}
+            <Strut size={16} />
+            <HeadingSmall>Subsequent ID:</HeadingSmall>
+            {id}
+        </View>
+    );
+};
+
+<UniqueIDProvider mockOnFirstRender={true}>
+    {children}
+</UniqueIDProvider>
+```
+
+### Scoped
+
+`UniqueIDProvider` ensures every identifier factory is unique using a unique number for each one. However, this isn't very readable when wanting to differentiate the types of things using unique identifiers. If we want to, we can provide a `scope` prop that adds some text to each identifier provided. This can be useful for providing some quick at-a-glance component identification to identifiers when there are multiple providers.
+
+```jsx
+const {Body, HeadingSmall, BodyMonospace} = require("@khanacademy/wonder-blocks-typography");
+const {Spring, Strut} = require("@khanacademy/wonder-blocks-layout");
+
+const children = ({id}) => (
+    <View>
+        <Body>
+            The id returned for "my-identifier": {id("my-identifier")}
```

Comment:
If I want multiple identifiers within `children` should I actually use multiple `UniqueIDProviders` and split up `children` into separate subtrees?

Reply:
I've updated the docs with more detail.